File: crates/kairos-soak/src/mcp.rs
//! MCP session driver over streamable HTTP (KAIROS-S-0006 transport,
//! mirroring `kairos-server/examples/e2e_golden_path.rs`): JSON-RPC over
//! POST `/mcp` with `Accept: application/json, text/event-stream`, the
//! `Mcp-Session-Id` header from `initialize`, SSE-framed responses, and a
//! best-effort DELETE at the end so hours-scale runs do not accumulate
//! server-side sessions.

use serde_json::{Value, json};

/// Extract the JSON-RPC message from a streamable-HTTP body (plain JSON
/// or SSE `data:` framing).
fn rpc_message(body: &str) -> Result<Value, String> {
    if let Ok(value) = serde_json::from_str::<Value>(body)
        && value.get("jsonrpc").is_some()
    {
        return Ok(value);
    }
    for line in body.lines() {
        if let Some(data) = line.strip_prefix("data:")
            && let Ok(value) = serde_json::from_str::<Value>(data.trim())
            && value.get("jsonrpc").is_some()
        {
            return Ok(value);
        }
    }
    Err(format!("no JSON-RPC message in response body: {body:?}"))
}

/// One JSON-RPC POST to `/mcp`; returns `(status, session header, body)`.
async fn post(
    http: &reqwest::Client,
    base_url: &str,
    tenant: &str,
    token: &str,
    session: Option<&str>,
    body: Value,
) -> Result<(u16, Option<String>, String), String> {
    let mut request = http
        .post(format!("{base_url}/mcp"))
        .bearer_auth(token)
        .header("accept", "application/json, text/event-stream")
        .header("x-tenant", tenant)
        .json(&body);
    if let Some(session) = session {
        request = request.header("mcp-session-id", session);
    }
    let response = request
        .send()
        .await
        .map_err(|e| format!("POST /mcp: {e}"))?;
    let status = response.status().as_u16();
    let session_id = response
        .headers()
        .get("mcp-session-id")
        .and_then(|v| v.to_str().ok())
        .map(str::to_string);
    let body = response
        .text()
        .await
        .map_err(|e| format!("/mcp body read: {e}"))?;
    Ok((status, session_id, body))
}

/// Call one tool inside an open session; errors carry the tool name.
#[allow(clippy::too_many_arguments)] // transport plumbing, called from one place
async fn call_tool(
    http: &reqwest::Client,
    base_url: &str,
    tenant: &str,
    token: &str,
    session: &str,
    id: i64,
    tool: &str,
    arguments: Value,
) -> Result<(), String> {
    let (status, _, body) = post(
        http,
        base_url,
        tenant,
        token,
        Some(session),
        json!({
            "jsonrpc": "2.0",
            "id": id,
            "method": "tools/call",
            "params": {"name": tool, "arguments": arguments},
        }),
    )
    .await?;
    if status != 200 {
        return Err(format!("{tool}: HTTP {status}: {body}"));
    }
    let message = rpc_message(&body).map_err(|e| format!("{tool}: {e}"))?;
    if message.get("error").is_some() {
        return Err(format!("{tool}: protocol error: {message}"));
    }
    if message["result"]["isError"].as_bool().unwrap_or(false) {
        return Err(format!("{tool}: tool error: {}", message["result"]));
    }
    Ok(())
}

/// Run one complete MCP session: initialize → notifications/initialized →
/// `whoami` → `board_items` → `get_item` → best-effort session DELETE.
pub async fn run_session(
    http: &reqwest::Client,
    base_url: &str,
    tenant: &str,
    token: &str,
    board_slug: &str,
    item_short_code: &str,
) -> Result<(), String> {
    let (status, session, body) = post(
        http,
        base_url,
        tenant,
        token,
        None,
        json!({
            "jsonrpc": "2.0",
            "id": 0,
            "method": "initialize",
            "params": {
                "protocolVersion": "2025-06-18",
                "capabilities": {},
                "clientInfo": {"name": "kairos-soak", "version": env!("CARGO_PKG_VERSION")},
            },
        }),
    )
    .await?;
    if status != 200 {
        return Err(format!("initialize: HTTP {status}: {body}"));
    }
    rpc_message(&body).map_err(|e| format!("initialize: {e}"))?;
    let session = session.ok_or("initialize returned no Mcp-Session-Id")?;

    let (status, _, body) = post(
        http,
        base_url,
        tenant,
        token,
        Some(&session),
        json!({"jsonrpc": "2.0", "method": "notifications/initialized"}),
    )
    .await?;
    if status != 202 {
        return Err(format!("notifications/initialized: HTTP {status}: {body}"));
    }

    call_tool(
        http,
        base_url,
        tenant,
        token,
        &session,
        1,
        "whoami",
        json!({}),
    )
    .await?;
    call_tool(
        http,
        base_url,
        tenant,
        token,
        &session,
        2,
        "board_items",
        json!({"board": board_slug}),
    )
    .await?;
    call_tool(
        http,
        base_url,
        tenant,
        token,
        &session,
        3,
        "get_item",
        json!({"short_code": item_short_code}),
    )
    .await?;

    // Best-effort session teardown (rmcp streamable HTTP supports DELETE);
    // an error here never fails the op.
    let _ = http
        .delete(format!("{base_url}/mcp"))
        .bearer_auth(token)
        .header("x-tenant", tenant)
        .header("mcp-session-id", &session)
        .send()
        .await;
    Ok(())
}

File: uat/surfaces/mcp.ts
// The MCP surface an agent persona holds: one streamable-HTTP session
// against /mcp (initialize → notifications/initialized → tools/call),
// the same wire the plugin's skills speak. Tool results come back as the
// text the skills read; small parsers below pick out the lines they key on.
import { runContext } from '../run/context';
import { recordSurface } from '../run/coverage';

function rpcMessage(body: string): any {
  try {
    const value = JSON.parse(body);
    if (value && value.jsonrpc) return value;
  } catch {
    /* SSE framing */
  }
  for (const line of body.split('\n')) {
    if (line.startsWith('data:')) {
      try {
        const value = JSON.parse(line.slice(5).trim());
        if (value && value.jsonrpc) return value;
      } catch {
        /* keep looking */
      }
    }
  }
  throw new Error(`no JSON-RPC message in MCP response: ${body.slice(0, 300)}`);
}

export class McpToolError extends Error {
  constructor(public readonly tool: string, public readonly text: string) {
    super(`${tool}: ${text}`);
  }
}

export class McpSession {
  private sessionId: string | undefined;
  private nextId = 1;

  constructor(
    private readonly token: string,
    private readonly clientName: string,
    private readonly server = runContext().server,
    private readonly tenant = runContext().tenant,
  ) {}

  private async post(body: unknown): Promise<{ status: number; session: string | null; text: string }> {
    const res = await fetch(`${this.server}/mcp`, {
      method: 'POST',
      headers: {
        authorization: `Bearer ${this.token}`,
        accept: 'application/json, text/event-stream',
        'content-type': 'application/json',
        'x-tenant': this.tenant,
        ...(this.sessionId ? { 'mcp-session-id': this.sessionId } : {}),
      },
      body: JSON.stringify(body),
    });
    return { status: res.status, session: res.headers.get('mcp-session-id'), text: await res.text() };
  }

  async initialize(): Promise<void> {
    if (this.sessionId) return;
    const init = await this.post({
      jsonrpc: '2.0',
      id: 0,
      method: 'initialize',
      params: {
        protocolVersion: '2025-06-18',
        capabilities: {},
        clientInfo: { name: `kairos-uat/${this.clientName}`, version: '0.0.0' },
      },
    });
    if (init.status !== 200) throw new Error(`MCP initialize HTTP ${init.status}: ${init.text}`);
    if (!init.session) throw new Error('MCP initialize returned no mcp-session-id');
    this.sessionId = init.session;
    const message = rpcMessage(init.text);
    if (message.result?.serverInfo?.name !== 'kairos') {
      throw new Error(`unexpected MCP serverInfo: ${init.text}`);
    }
    const notified = await this.post({ jsonrpc: '2.0', method: 'notifications/initialized' });
    if (notified.status !== 202) throw new Error(`MCP initialized notification HTTP ${notified.status}`);
  }

  /**
   * Call a tool and return its text. A tool-level error (`isError`) throws
   * `McpToolError` carrying the text, so a journey can assert on a refusal.
   */
  async call(name: string, args: Record<string, unknown> = {}): Promise<string> {
    recordSurface('mcp', name);
    await this.initialize();
    const res = await this.post({
      jsonrpc: '2.0',
      id: this.nextId++,
      method: 'tools/call',
      params: { name, arguments: args },
    });
    if (res.status !== 200) throw new Error(`${name} HTTP ${res.status}: ${res.text}`);
    const message = rpcMessage(res.text);
    if (message.error) throw new Error(`${name} protocol error: ${JSON.stringify(message.error)}`);
    const result = message.result ?? {};
    const text = (result.content ?? [])
      .filter((c: any) => c.type === 'text')
      .map((c: any) => c.text)
      .join('\n');
    if (result.isError) throw new McpToolError(name, text);
    return text;
  }

  /**
   * The tool names this deployment offers (`tools/list`). The coverage
   * gate asks the server rather than hard-coding a list, so a tool added
   * to the product shows up here the day it ships.
   */
  async listTools(): Promise<string[]> {
    await this.initialize();
    const res = await this.post({ jsonrpc: '2.0', id: this.nextId++, method: 'tools/list' });
    if (res.status !== 200) throw new Error(`tools/list HTTP ${res.status}: ${res.text}`);
    const message = rpcMessage(res.text);
    if (message.error) throw new Error(`tools/list error: ${JSON.stringify(message.error)}`);
    return ((message.result?.tools ?? []) as any[]).map((t) => t.name as string).sort();
  }

  /** `call` that returns the refusal text instead of throwing. */
  async refused(name: string, args: Record<string, unknown> = {}): Promise<string> {
    try {
      const text = await this.call(name, args);
      throw new Error(`${name} was expected to be refused but returned: ${text.slice(0, 200)}`);
    } catch (err) {
      if (err instanceof McpToolError) return err.text;
      throw err;
    }
  }
}

/** Short codes like DEMO-T-0042 found in tool text, in order, de-duplicated. */
export function shortCodes(text: string): string[] {
  return [...new Set(text.match(/\b[A-Z][A-Z0-9]*-[A-Z]-\d{4}\b/g) ?? [])];
}

/** The `key: value` line a tool prints (e.g. `repository: payments-api (owner: platform)`). */
export function field(text: string, key: string): string | undefined {
  const re = new RegExp(`^\\s*${key}:\\s*(.+)$`, 'm');
  return text.match(re)?.[1]?.trim();
}
